Add country and insitute to user model

garry329 opened the pull request:
Fixes #4 

#### Changes I have made
- added country and educational institute field to user model
- added country_select gem to get country list
- added country and educational institute form fields to edit profile form and show them in user profile
- refactoring code in logix controller
- formatting some code in view templates

### Screenshots of the changes -

![screenshot 2019-03-03 at 2 37 51 pm](https://user-images.githubusercontent.com/20072704/53693141-0cd8df00-3dc2-11e9-93aa-1adbefeb049d.png)
![screenshot 2019-03-03 at 2 37 44 pm](https://user-images.githubusercontent.com/20072704/53693143-106c6600-3dc2-11e9-8ab1-4361899d51f1.png)

tachyons commented on app/controllers/users/logix_controller.rb:
```diff
@@ -1,28 +1,44 @@
   end
+
+  def profile_params
+    params.require(:user).permit(:name,:profile_picture,:country,:educational_institute)
```
indentation

tachyons commented on app/views/users/logix/edit.html.erb:
```diff
@@ -1,46 +1,40 @@
+
+        <div class="field form-group">
+          <%= f.label :educational_institute %><br/>
+          <%= f.text_field :educational_institute,{ class: "form form-group form-control" } %>
```
Here we should list list of educational institutions already available in our database as suggestions to avoid common typos. You can fetch them from `User.distinct.pluck(:educational_institute)`

tachyons replied:
It should look like this http://twitter.github.io/typeahead.js/examples/

garry329 replied:
so should i use typeahead.js in the project to achieve this?

tachyons replied:
Not necessary, I just gave an example. You can use any other library that does this

garry329 replied:
Ok, got it!

garry329 replied:
So, I have added autocomplete to the educational institute option.I went with the typeahead.js library that you suggested.Please review the code and let me know if something needs to be improved.

![autocomplete](https://user-images.githubusercontent.com/20072704/53696119-f6df1480-3de9-11e9-974f-86316a70291d.gif)

garry329 replied:
@tachyons , can you please check the changes i have made?

tachyons replied:
Can you check if avatar is persisting, I see there was a custom code to handle it

tachyons commented on app/views/users/logix/profile.html.erb:
```diff
@@ -1,32 +1,29 @@
+          <p><strong>Member since: </strong><%= ((Time.now.to_i - @profile.created_at.to_i) / 1.day) %> days</p>
+          <p><strong>Educational Institution: </strong><%= @profile.educational_institute %></p>
+          <p><strong>Country: </strong><%= @profile.country_name %></p>
+          <% if user_signed_in? and current_user.id == @profile.id %>
```
You are setting @profile as current user, right? So how this logic is supposed to work ?

garry329 replied:
Ok, i see where i went wrong.I have fixed this in my latest commit.

garry329 replied:
Can you please check that?

tachyons commented on app/controllers/users/logix_controller.rb:
```diff
@@ -1,28 +1,49 @@
+  def typeahead_educational_institute
+    query = params[:query]
+    educational_institute_list = User.where("educational_institute LIKE ?", "%#{query}%").uniq { |user| user.educational_institute }
+                                     .pluck(:educational_institute)
```
This query can be expensive, uniq is an array operation, not a db operation. do uniqueness check in database query itself

garry329 replied:
@tachyons, so i have fixed this by replacing uniq with distinct in my latest commit.

tachyons commented:
Looks good so far, Can we have tests as well ?

garry329 commented on app/controllers/users/logix_controller.rb:
```diff
@@ -18,6 +18,12 @@ def edit
 
+  def typeahead_educational_institute
+    query = params[:query]
+    educational_institute_list = User.where("educational_institute LIKE ?", "%#{query}%").map{|user| {name:user.educational_institute}}
```
@tachyons i think that would have happened if a string had been passed as a sql fragment in something like
`User.where("name = '#{params[:name]'")`
but i am using a template and passing params to it like
`User.where(["name = ?", "#{params[:name]}"])`.what

garry329 commented:
Ok, cool.

garry329 commented:
@tachyons can you please tell me what all kinds of tests should i write?

garry329 commented:
@tachyons i have added the testcases, and also :
- removed the  assignment_submissions.yml as it has been removed from schema and was causing issues in running of tests.
- removed the mentor field from assignment.yml which wasn't part of assignment model and was causing an error in test

Please review the changes and let me know if anything more needs to be done.

tachyons commented:
We decided to chose rspec + factory not pair for testing. Can you please port your code from ActionDispatch integration test to rspec integration test? Because consistency is important

garry329 commented:
Ok, Sure.I am on it.

garry329 commented:
@tachyons, I have added the rspec testcases.Please review the code and let me know if anything else is to be done.

tachyons commented on app/views/users/logix/edit.html.erb:
```diff
@@ -1,46 +1,57 @@
-
-
+<link href="/css/typeahead.css" rel="stylesheet">
```
Can we use minimised version if we are using css directly

garry329 replied:
i have added the minified css.

tachyons commented on spec/controllers/users_logix_controller_spec.rb:
```diff
@@ -0,0 +1,39 @@
+    sign_in @user
+  end
+
+  it  'should get user projects' do
```
space ✂️

garry329 replied:
fixed

tachyons commented on spec/controllers/users_logix_controller_spec.rb:
```diff
@@ -0,0 +1,39 @@
+  end
+
+  it 'should update user profile' do
+    patch profile_update_path(@user), params:{id:@user.id,user:{"name"=>"Jd", "country"=>"IN", "educational_institute"=>"MAIT"}}
```
new hash syntax

garry329 replied:
fixed

tachyons commented on test/controllers/users_logix_controller_test.rb:
```diff
@@ -0,0 +1,48 @@
+require 'test_helper'
+
+class UsersLogixControllerTest < ActionDispatch::IntegrationTest
```
✂️

garry329 replied:
removed this file as its no longer required

tachyons commented on test/fixtures/assignments.yml:
```diff
@@ -2,12 +2,10 @@
 
```
✂️ use factory bot

garry329 replied:
@tachyons , these were changes from the integration tests which are longer being used.So i have removed these changes from this PR.

tachyons commented on test/fixtures/users.yml:
```diff
@@ -4,8 +4,12 @@
 # model remove the '{}' from the fixture names and add the columns immediately
 # below each fixture, per the syntax in the comments below
 #
-one: {}
```
use factory bot

garry329 replied:
@tachyons , i have removed these changes from this PR(no longer used)

garry329 commented:
@tachyons , Please check the chnages i have made as per your suggestions.

tachyons commented:
Thanks

garry329 commented:
Thank you @tachyons , now i'll move on to the search feature.
